research/picai/fl_nnunet/nnunet_client.py

```diff
@@ -0,0 +1,607 @@
+        self.num_spatial_dims = get_num_spatial_dims(nnunet_config)
+
+        # nnUNetDataloaders store their datasets under the self.data attribute
+        self.dataset: nnUNetDataset = self.nnunet_dataloader.generator._data
+        super().__init__(dataset=self.dataset, batch_size=self.nnunet_dataloader.generator.batch_size)
+
+        # nnunet dataloaders are infinite by default so we have to track steps to stop iteration
+        self.current_step = 0
+        self.infinite = infinite
+        self.len = len(self.dataset)
+
+    def __next__(self) -> Tuple[torch.Tensor, Union[torch.Tensor, Dict[str, torch.Tensor]]]:
+        if not self.infinite and self.current_step == self.len:
+            self.reset()
+            raise StopIteration
+        else:
+            self.current_step += 1
+            batch = next(self.nnunet_dataloader)  # This returns a dictionary
+            # Note: When deep supervision is on, target is a list of segmentations at various scales
+            # nnUNet has a wrapper for loss functions to enable deep supervision
+            inputs: torch.Tensor = batch["data"]
+            if isinstance(batch["target"], list):
+                target_dict = convert_ds_list_to_dict(batch["target"], self.num_spatial_dims)
+                return inputs, target_dict
+            elif isinstance(batch["target"], torch.Tensor):
+                target_tensor: torch.Tensor = batch["target"]
+                return inputs, target_tensor
+            else:
+                raise TypeError(
+                    "Was expecting the target generated by the nnunet dataloader to be a list or a torch.Tensor"
+                )
+
+    def __len__(self) -> int:
+        # Should return num_samples // batch_size
+        # nnUNetDataloaders are 'infinite' meaning they randomly sample batches
+        # from the dataset. This makes the distinction between
+        # num_samples // batch_size and ceil(num_samples/batch_size)
+        # meaningless. We will abritrarily use the later
+        num_samples = len(self.dataset)
+        batch_size = self.nnunet_dataloader.generator.batch_size
+        return int(ceil(num_samples / batch_size))
+
+    def reset(self) -> None:
+        self.current_step = 0
+
+    def __iter__(self) -> DataLoader:  # type: ignore
+        # mypy gets angry that the return type
+        return self
+
+
+class Module2LossWrapper(_Loss):
+    """Converts a nn.Module subclass to a _Loss subclass"""
+
+    def __init__(self, loss: nn.Module, **kwargs: Any) -> None:
+        super().__init__(**kwargs)
+        self.loss = loss
+
+    def forward(self, pred: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
+        return self.loss(pred, target)
+
+
+class nnUNetClient(BasicClient):
+    def __init__(
+        self,
+        dataset_id: int,
+        data_identifier: Optional[str],
+        plans_identifier: Optional[str],
+        always_preprocess: bool = True,
+        **kwargs: Any,
```

Comment:
Just didn't want to clutter method with too many arguments, but in this instance theres not so many so will explicitly pass all the arguments

Reply:
Sounds good!

Reply:
P.S. This led to me realising the fold should be a client argument not a server argument. In instance where there are multiple clients, we don't need to enforce that they are training on the same fold, especially if the datasets are different. Leave it up to the user to ensure that the validation set for one client is not in the training set of another. Technically this could be an issue outside of nnunet anyways so no sense it restricting all nnunet clients to use the same fold